### DISCUSSION TITLE
Help Needed with Optimizing a Switching Cost Function in CasADi

### DISCUSSION THREAD

**Discussion by ZionLIU-hit** [casadi/casadi#3554] [Q&A]
Hello！
I am currently working on an optimization problem using CasADi where I need to switch between different cost functions based on an optimization variable. To achieve this, I've utilized the` if_else` function to construct the cost function. However, I am encountering an issue where this particular variable does not seem to be optimized during the process. 
To investigate further, I calculated the gradient of the cost function with respect to ts using CasADi's `gradient` function and found it to be zero. I am seeking advice on how to address this issue. Is there a particular approach or consideration in CasADi for handling such dynamic switching in cost functions? Thank you in advance!

Here's a snippet of my code for reference:

```
for k = 1:para.N

    ts = U(n_u+1:end,k);  % ts is the switching variable to be optimized
    phi = if_else(tau <= ts, 1, 0);

    obj =  obj + phi*(stq-P(n_q+1:n_q*2))'*para.Qt*(stq-P(n_q+1:n_q*2))...
               + (1-phi)*(st-P(n_q*2+1:n_q*2+n_ee))'*para.Qp*(st-P(n_q*2+1:n_q*2+n_ee))...
               + con'*para.R*con ;
end

grad_obj_ts = jacobian(obj, ts)
>> val = 0
```

**Comment by jgillis** [comment]
What is tau in your case? Can you post code snippet that is runnable? It may be needed to smoothen your switch statement, e.g. with tanh or a clipped sine.
